rasa/nlu/featurizers/sparse_featurizer/lexical_syntactic_featurizer.py

```diff
@@ -153,157 +372,130 @@ def _map_features_to_indices(
 
         return feature_to_idx_dict
 
-    @staticmethod
-    def _build_feature_vocabulary(features: List[List[Dict[Text, Any]]]) -> OrderedDict:
-        feature_vocabulary = defaultdict(set)
-
-        for sentence_features in features:
-            for token_features in sentence_features:
-                for feature_name, feature_value in token_features.items():
-                    feature_vocabulary[feature_name].add(feature_value)
+    def process(self, messages: List[Message]) -> List[Message]:
+        """Featurizes all given messages in-place.
 
-        # sort items to ensure same order every time (for tests)
-        ordered_feature_vocabulary = OrderedDict(sorted(feature_vocabulary.items()))
-
-        return ordered_feature_vocabulary
+        Returns:
+          the given list of messages which have been modified in-place
+        """
+        for message in messages:
+            self._process_message(message)
+        return messages
 
-    def _create_sparse_features(self, message: Message) -> None:
-        """Convert incoming messages into sparse features."""
-        import scipy.sparse
+    def _process_message(self, message: Message) -> None:
+        """Featurize the given message in-place.
 
+        Args:
+          message: a message to be featurized
+        """
+        if not self._feature_to_idx_dict:
+            rasa.shared.utils.io.raise_warning(
+                f"The {self.__class__.__name__} {self.identifier} has not been "
+                f"trained properly yet. "
+                f"Continuing without adding features from this featurizer."
+            )
+            return
         tokens = message.get(TOKENS_NAMES[TEXT])
-        # this check is required because there might be training data examples without
-        # TEXT, e.g., `Message("", {action_name: "action_listen"})`
         if tokens:
-
-            sentence_features = self._tokens_to_features(tokens)
-            one_hot_seq_feature_vector = self._features_to_one_hot(sentence_features)
-
-            sequence_features = scipy.sparse.coo_matrix(one_hot_seq_feature_vector)
-
-            final_sequence_features = Features(
-                sequence_features,
-                FEATURE_TYPE_SEQUENCE,
-                TEXT,
-                self.component_config[FEATURIZER_CLASS_ALIAS],
+            sentence_features = self._map_tokens_to_raw_features(tokens)
+            sparse_matrix = self._map_raw_features_to_indices(sentence_features)
+            self.add_features_to_message(
+                # FIXME: create sentence feature and make `sentence` non optional
```

Comment:
should we create an issue for this?

Reply:
done -> #9574

Reply:
Thanks! added it to the Research project 👍🏻